File: src/cli_server/weather_py_server/weather_cli.py
import logging

import subprocess
from fastapi import FastAPI, HTTPException, Response, status
from pydantic import BaseModel

import os
import ast
import re

import sys
from pathlib import Path
import asyncio

class CodeRequest(BaseModel):
    language: str
    code: str

app = FastAPI()

@app.post("/execute")
async def execute_code(request: CodeRequest, response: Response):
    # Mapping languages to execution commands
    # Add more as needed (e.g., 'javascript': ['node', '-e'])
    commands = {
        "python": ["python3", "-c"],
        "bash": ["bash", "-c"]
    }

    # raise HTTPException(status_code=438, detail=f"Random test error for debugging purposes.")

    try:
        validate_code_integrity(request.language.lower(), request.code)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=f"Invalid code grammar: {str(e)}")

    try:
        imports = "from weather import get_forecast, get_alerts\n"
        full_code = imports + request.code
        script_dir = Path(__file__).resolve().parent

        process = await asyncio.create_subprocess_exec(
            *commands[request.language.lower()],
            full_code,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            cwd=script_dir
        )

        try:
            # Wait for completion with a timeout
            stdout, stderr = await asyncio.wait_for(process.communicate(), timeout=10.0)
        except asyncio.TimeoutError:
            process.kill()
            raise HTTPException(status_code=status.HTTP_408_REQUEST_TIMEOUT, detail="Execution timed out.")

        if process.returncode != 0:
            error_msg = stderr.decode().strip()
            logging.error(f"Execution error:\n{error_msg}")
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=f"Execution error: {error_msg}")
        
        response.status_code = status.HTTP_200_OK
        return {
            "stdout": stdout.decode()
        }
    
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=f"Internal Error: {str(e)}")

def validate_code_integrity(language: str, code: str):
    # 1. Language Label Verification (Heuristics)
    markers = {
        "python": [r"def\s+\w+\(", r"import\s+\w+", r"print\(", r"self\."],
        "bash": [r"echo\s+", r"\$\(", r"if\s+\[", r"sudo\s+", r"apt-get"]
    }
    
    if language in markers:
        # Check if at least one marker exists for the labeled language
        if not any(re.search(pattern, code) for pattern in markers[language]):
            raise ValueError(f"Code content does not appear to match the '{language}' label.")

    # 2. Syntax Validation (Dry Run)
    if language == "python":
        try:
            compile(code, "<string>", "exec")
        except SyntaxError as e:
            raise ValueError(f"Python Syntax Error: {e.msg} at line {e.lineno}")
            
    elif language == "bash":
        check = subprocess.run(["bash", "-n", "-c", code], capture_output=True, text=True)
        if check.returncode != 0:
            raise ValueError(f"Bash Syntax Error: {check.stderr.strip()}")

if __name__ == "__main__":
    import uvicorn
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 8000
    uvicorn.run(app, host="0.0.0.0", port=port)

File: src_ts/cli_server/weather_ts_server/weather_cli.ts
import process from "node:process";
import { spawn } from "node:child_process";
import Fastify from "fastify";
import { z } from "zod";

function logInfo(message: string) {
  process.stdout.write(`INFO:    [weather_cli] ${message}\n`);
}

function logError(message: string) {
  process.stderr.write(`ERROR:   [weather_cli] ${message}\n`);
}

const CodeRequestSchema = z.object({
  language: z.string(),
  code: z.string(),
});

function validateCodeIntegrity(language: string, code: string) {
  const lang = language.toLowerCase();
  const markers: Record<string, RegExp[]> = {
    javascript: [/console\.log\(/, /await\s+/, /function\s+/, /=>/],
    node: [/console\.log\(/, /await\s+/, /require\(|from\s+["']/, /import\s+/],
  };

  if (markers[lang]) {
    if (!markers[lang].some((re) => re.test(code))) {
      throw new Error(`Code content does not appear to match the '${language}' label.`);
    }
  }
}

const app = Fastify({ logger: false });

app.get("/openapi.json", async () => {
  // Orchestrator readiness probe parity with FastAPI
  return { openapi: "3.0.0", info: { title: "weather-ts-server", version: "0.1.0" }, paths: {} };
});

app.post("/execute", async (request, reply) => {
  const parsed = CodeRequestSchema.safeParse(request.body);
  if (!parsed.success) {
    reply.status(422);
    return { detail: `Invalid request: ${parsed.error.message}` };
  }

  const { language, code } = parsed.data;
  const lang = language.toLowerCase();

  try {
    validateCodeIntegrity(lang, code);
  } catch (e: any) {
    reply.status(422);
    return { detail: `Invalid code grammar: ${e?.message ?? String(e)}` };
  }

  if (lang !== "javascript" && lang !== "node") {
    reply.status(422);
    return { detail: `Unsupported language: ${language}. Use 'javascript' (node).` };
  }

  const prelude = `
const { getForecast, getAlerts } = require("./weather");
(async () => {
${code}
})().catch((e) => { console.error(e && e.stack ? e.stack : String(e)); process.exitCode = 1; });
`.trim();

  const scriptDir = __dirname;
  const timeoutMs = 10_000;

  let killed = false;
  const child = spawn("node", ["-e", prelude], { cwd: scriptDir });
  const done = new Promise<{ stdout: string; stderr: string; code: number | null }>((resolve) => {
    let stdout = "";
    let stderr = "";
    child.stdout.on("data", (d) => (stdout += d.toString()));
    child.stderr.on("data", (d) => (stderr += d.toString()));
    child.on("close", (code) => resolve({ stdout, stderr, code }));
  });

  const timer = setTimeout(() => {
    killed = true;
    child.kill("SIGKILL");
  }, timeoutMs);

  const result = await done.finally(() => clearTimeout(timer));
  if (killed) {
    reply.status(408);
    return { detail: "Execution timed out." };
  }
  if (result.code !== 0) {
    logError(`Execution error:\n${result.stderr.trim()}`);
    reply.status(400);
    return { detail: `Execution error: ${result.stderr.trim()}` };
  }

  reply.status(200);
  return { stdout: result.stdout };
});

if (require.main === module) {
  const port = Number(process.argv[2] ?? "8000");
  app.listen({ host: "0.0.0.0", port }).then(() => logInfo(`Listening on 0.0.0.0:${port}`));
}

